internal/providers/vmware/vmware_amd64.go

```diff
@@ -50,12 +52,70 @@ func FetchConfig(f *resource.Fetcher) (types.Config, report.Report, error) {
 }
 
 func fetchRawConfig(f *resource.Fetcher) (config, error) {
+	var data string
+	var encoding string
+	var err error
+
+	data, err = getVariable(f, "ignition.config.data")
+	if err == nil && data != "" {
+		encoding, err = getVariable(f, "ignition.config.data.encoding")
+	} else {
+		data, err = getVariable(f, "coreos.config.data")
+		if err == nil && data != "" {
+			encoding, err = getVariable(f, "coreos.config.data.encoding")
+		}
+	}
+
+	// Return an error for an empty config to simplify falling back to the URL
+	// Do not check against err from "encoding" because leaving it empty is ok
+	if data == "" {
+		f.Logger.Debug("failed to fetch config")
+		return config{}, providers.ErrNoProvider
+	}
+
+	return config{
+		data:     data,
+		encoding: encoding,
+	}, nil
+}
+
+func fetchUrlConfig(f *resource.Fetcher) (types.Config, report.Report, error) {
+	rawUrl, err := getVariable(f, "ignition.config.url")
+	if err != nil || rawUrl == "" {
+		rawUrl, err = getVariable(f, "coreos.config.url")
+	}
+
+	if err != nil || rawUrl == "" {
+		f.Logger.Info("no config URL provided")
+		return types.Config{}, report.Report{}, providers.ErrNoProvider
+	}
+
+	f.Logger.Debug("found url: %q", rawUrl)
+
+	url, err := url.Parse(rawUrl)
+	if err != nil {
+		f.Logger.Err("failed to parse url: %v", err)
+		return types.Config{}, report.Report{}, err
+	}
+
+	if url == nil {
+		return types.Config{}, report.Report{}, providers.ErrNoProvider
+	}
+
+	data, err := f.FetchToBuffer(*url, resource.FetchOptions{
+		Headers: resource.ConfigHeaders,
+	})
+	if err != nil {
+		return types.Config{}, report.Report{}, err
+	}
+
+	return util.ParseConfig(f.Logger, data)
+}
+
+func getVariable(f *resource.Fetcher, key string) (string, error) {
 	info := rpcvmx.NewConfig()
 
 	var ovfData string
-	var ovfEncoding string
-	ovfDataKey := "coreos.config.data"
-	ovfEncodingKey := "coreos.config.data.encoding"
 
 	ovfEnv, err := info.String("ovfenv", "")
```

Comment:
FYI, this now fetches the whole of `ovfenv` for every variable lookup. However, since only a few variables are looked up I don't suppose this matters.

Reply:
I think it's better to keep it simple instead of optimizing this little overhead away.